whylog/rulesbase/__init__.py

```diff
@@ -1,6 +1,144 @@
+        parsers_dict = {}
+        for parser in parsers:
+            parsers_dict[parser.get('name')] = Parser(
+                parser.get('name'), 
+                parser.get('regex'), 
+                self._get_params_for_parser(parser)
+            )
+        self._parsers = parsers_dict
+
+    def _get_causes_for_rule(self, rule_dict):
+        if 'cause' in rule_dict:
+            return [rule_dict.get('cause')]
+        else:
+            return rule_dict.get('causes')
+
+    def _get_constraints_for_rule(self, rules_dict):
+        if 'constraints' in rules_dict:
+            return rules_dict.get('constraints')
+        else:
+            return None
+
+    def _get_type_for_rule(self, rules_dict):
+        if 'type' in rules_dict:
+            return rules_dict.get('type')
+        else:
+            return None
+
+    def _get_cause_constraints_for_rule(self, rules_dict):
+        if 'cause_constraints' in rules_dict:
+            return rules_dict.get('cause_constraints')
+        else:
+            return None
+
+    def _load_rules(self):
+        rules = self._load_config_from_file(self._rules_path)
+        rules_dict = defaultdict(list)
+        for rule in rules:
+            rules_dict[rule.get('effect')].append(Rule(
+                self._get_causes_for_rule(rule), 
+                rule.get('effect'), 
+                self._get_constraints_for_rule(rule), 
+                self._get_type_for_rule(rule), 
+                'negative' in rules_dict,
+                self._get_cause_constraints_for_rule(rule)
+                )
+            )
+        self._rules = rules_dict
+
+    def get_parsers(self):
+        return self._parsers
+
+    def get_rule(self, effect_name):
+        if effect_name in self._rules:
+            return self._rules.get(effect_name)
+        else:
+            return None
+
+class Parser(object):
+    def __init__(self, name, regex, params):
+        self._name = name
+        self._regex_str = regex
+        self._regex = re.compile(regex)
+        self._params = params
+
+    def get_name(self):
+        return self._name 
+    
+    def get_regex_str(self):
+        return self._regex_str
+
+    def get_regex(self):
+        return self._regex
+
+    def get_params(self):
+        return self._params
+
+    def match(self, line_str):
+        return self._regex.match(line_str)
+
+    def get_regex_params_from_line(self, line_str):
+        pass
+
+class Rule(object):
+    def __init__(self, causes, effect, constraints, 
+        rule_type, negative, cause_constraints):
+        self._causes = causes
+        self._effect = effect
+        self._constraints = constraints
+        self._rule_type = rule_type
+        self._negative = negative
+        self._cause_constraints = cause_constraints
+
+    def get_causes(self):
+        return self._causes
+
+    def get_effect(self):
+        return self._effect
+
+    def get_constraints(self):
```

Comment:
the caller does not want to "get" constraints or check if the rule is negative. The caller will want to check if the condition of the rule is satisfied: if cause match and work together in the specified way (AND/OR). If if the rule is inverted, then caller wants to know the opposite. Every caller of methods of this class does not want to write code like
```python
for cause in rule.get_causes():
    if rule.get_negative and not cause.is_satisfied(current_situation) or not rule.get_negative and cause.is_satisfied(current_situation):
        continue # etc
    return []
return rule.get_effects()
```
but rather
```python
effects = rule.matched_effects(current_situation)
```